### DISCUSSION TITLE
It seems like sse events are processed synchronously, is this the case?

### DISCUSSION THREAD

**Discussion by Ancient-Dragon** [enisdenjo/graphql-sse#56] [General]
We have recently been doing some performance testing and we found that our server would send lots of events in quick succession and by using the debugMessage + other logs we noticed that it was waiting for the processing of that subscription handler to be completed before processing the next event (didn't even have to be the same subscription). 

Just wanting to confirm whether that is the case? and if so is it possible to make them async (especially when they're not the same subscription)?

**Comment by enisdenjo** [comment]
JS is not concurrent by nature, it is a single-threaded language being able to only execute one task at a time. But I guess you already knew that, so I might be making a wrong assumption here - can you therefore elaborate a bit more? Or even better, create a repro showcasing the behaviour?

**Reply by Ancient-Dragon** [reply]
I guess what I'm asking is if it calls the callback functions via promises? So if I write a subscription handler which calls an api or something would it wait for that to finish processing before consuming another sse event? Or would it consume another event while that method is "paused" waiting for the response from the http request

**Reply by enisdenjo** [reply]
All callbacks are promises, and they are run synchronous to the _current_ subscription because its values are necessary for further execution. However, the request handler itself is a promise too; meaning, each new subscription request is a promise too.

**Reply by Ancient-Dragon** [reply]
We were also seeing this across subscriptions though, one subscription was holding up an event for another one. So our backend service would emit, two sse events for different subscription is quick succession (under ~100ms) and then the frontend would process the first one (which was taking ~1s) and the second sse wasn't even logged until after all of that code had completed.

The other question I was going to ask at some point, was is there anyway to handle subscriptions in a web worker?

**Reply by enisdenjo** [reply]
Well if the processing of the first event took up all of main thread's resources (was not suspended in a promise), then it makes sense that the other event only happened afterwards.

About the web workers, I think `fetch` can run within - so I guess you can run graphql-sse there too. You can also, alternatively, only do the processing of events in web workers but still have graphql-sse in the main thread.

P.S. the client code is also "promisified".

**Reply by danstarns** [reply]
Heya, it's worth chiming in here that I found similar issues when the server would fire two events at the same time, with the same timestamp, one of which was dropped. Might be related - see you in 2 years ;)

**Reply by enisdenjo** [reply]
The issue for https://github.com/enisdenjo/graphql-sse/discussions/56#discussioncomment-13435699 was actually in urql, graphql-sse emits all events.
